marimo/_save/save.py

```diff
@@ -0,0 +1,161 @@
+        self._entered_trace = False
+        self._old_trace: Optional[TraceFunction] = None
+
+    def __enter__(self) -> Self:
+        sys.settrace(lambda *_args, **_keys: None)
+        frame = sys._getframe(1)
+        # Attempt to hold on to the previous trace.
+        self._old_trace = frame.f_trace
+        # Setting the frametrace, will cause the function to be run on _every_
+        # single context call until the trace is cleared.
+        frame.f_trace = self.trace
+        return self
+
+    def trace(
+        self, _frame: FrameType, _event: str, _arg: Any
+    ) -> Union[TraceFunction | None]:
+        # General flow is as follows:
+        #   1) Follow the stack trace backwards to the first instance of a
+        # "<module>" function call, which corresponds to a cell level block.
+        #   2) Run static analysis to determine whether the call meets our
+        # criteria. The procedure is a little brittle as such, certain contexts
+        # are not allow (e.g. called within a function or a loop).
+        #  3) Hash the execution and lookup the cache, and return!
+        #  otherwise) Set _skipped such that the block continues to execute.
+
+        self._entered_trace = True
+        stack = traceback.extract_stack()
+        if not self._skipped:
+            return self._old_trace
+
+        # This only executes on the first line of code in the block. If the
+        # cache is hit, the block terminates early with a SkipWithBlock
+        # exception, if the block is not hit, self._skipped is set to False,
+        # causing this function to terminate before reaching this block.
+        for i, frame in enumerate(stack[::-1]):
+            _filename, lineno, function_name, _code = frame
+            if function_name == "<module>":
+                ctx = get_context()
+                assert (
+                    ctx.execution_context is not None
+                ), "Could not resolve context for cache."
+                graph = ctx.graph
+                cell_id = ctx.execution_context.cell_id
+                pre_module, save_module = ExtractWithBlock(lineno - 1).visit(
+                    ast.parse(ctx.graph.cells[cell_id].code).body
+                )
+
+                self._cache = hash_context(
+                    save_module,
+                    graph,
+                    cell_id,
+                    loader=self._loader,
+                    context=pre_module,
+                )
+                if self._cache.hit:
+                    raise SkipWithBlock()
+
+                self.cache_type = self._cache
+                self._skipped = False
+                return self._old_trace
+            elif i > 1:
+                raise CacheException(
+                    "persistent_cache must be invoked from cell level "
+                    "(cannot be in a function or class)"
+                )
+        raise CacheException("persistent_cache could not resolve block")
```

Comment:
Are there valid cases in which this would be hit, or would it be considered a bug in marimo if this exception were raised?

If latter, the error message could be "persistent_cache could not resolve block — this is a bug in marimo. Please file an issue at https://github.com/marimo-team/marimo/issues"